os/pm/pm_timer/pm_timer.h

```diff
@@ -0,0 +1,117 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name NuttX nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+#ifndef __PM_TIMER_H
+#define __PM_TIMER_H
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <assert.h>
+#include <debug.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <queue.h>
+
+#define PM_TIMER_FAIL     -1
+#define PM_TIMER_SUCCESS   0
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Public Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Public Variables
+ ************************************************************************/
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C" {
+#else
+#define EXTERN extern
+#endif
+
+/* The g_pm_activeList data structure is a singly linked list ordered by
+ * pm wakeup timer expiration time.
+ */
+
+extern sq_queue_t g_pmTimer_activeList;
+
+/* This is the number of free, pre-allocated pm wakeup timer structures in the
+ * g_pmTimer_list. 
+ */
+
+extern uint16_t g_pmTimer_nfree;
+
+/* g_pmTimer_list is a list of pre-allocated pm timers. The number of pm timers
+ * in the list is a configuration item.
+ */
+
+extern pm_wakeup_timer_t g_pmTimer_list[CONFIG_PM_MAX_WAKEUP_TIMER];
```

Comment:
Could you let me know why you want to use pre-allocated and fixed array?

Reply:
Here are the following reasons:
1. code complexity is very less. It is easy to find the timer node using index. High efficiency(faster) and less cases to handle.
2. If we use dynamically allocated timer, we need a way to find the different timer whenever an apps call a specific timer, we will need an indexing method because we should not provide the address to the app. Apart from that decrease in efficiency but good memory usage.
3. If we dont need a lot of timers , then i think it is more efficient to just allocate around 10 timer statically ?